### PR TITLE
Add QSD installation test

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2150238
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=qsd_install --iothread_scheme=roundrobin --nr_iothreads=2 --platform=x86_64 --guestname=RHEL.8.8.0,Win2019  --imageformat=raw,qcow2 --machines=q35,i440fx --customsparams="cpu_model_flags = 'vmx,+kvm_pv_unhalt'" --firmware=default_bios,ovmf --netdst=virbr0 --clone=no

01-Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install Finshed 2022-12-01 03:11:08 2022-12-01 03:31:27 PASS 1219.545143
02-Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 03:31:28 2022-12-01 03:51:54 PASS 1225.612273
03-Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install Finshed 2022-12-01 03:51:55 2022-12-01 04:31:21 PASS 2366.005241
04-Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 04:31:22 2022-12-01 05:10:08 PASS 2326.379568
05-Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install Finshed 2022-12-01 05:10:09 2022-12-01 05:30:33 PASS 1223.830057
06-Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 05:30:34 2022-12-01 06:00:26 PASS 1792.606651
07-Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install Finshed 2022-12-01 06:00:27 2022-12-01 06:37:58 PASS 2250.824432
08-Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 06:37:59 2022-12-01 07:12:58 PASS 2098.819222
09-Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 07:12:59 2022-12-01 07:34:23 PASS 1284.302634
10-Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 07:34:24 2022-12-01 08:15:10 PASS 2446.502636
11-Host_RHEL.m9.u2.ovmf.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 08:15:11 2022-12-01 08:36:20 PASS 1268.579062
12-Host_RHEL.m9.u2.ovmf.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35 Finshed 2022-12-01 08:36:21 2022-12-01 09:14:06 PASS 2265.48448

**Comment by qingwangrh** [comment]
Hi, @YongxueHong @zhenyzha @XueqiangWei @menli820 Please help to review,thanks.

File: qemu/tests/cfg/qsd_install.cfg [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    unattended_delivery_method = cdrom
+    medium = cdrom
+    cdroms += " unattended"
+    guest_cmd = lsblk
+    cd_format_fixed = ahci
+    drive_index_unattended = 1
+    drive_index_cd1 = 2
+
+    ovmf:
+        only q35
```
Comment (XueqiangWei) [inline qemu/tests/cfg/qsd_install.cfg:32]: Do we need to consider  aarch64 platform?

File: qemu/tests/qsd_install.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        logger.debug("Reboot...")
+        params["cdroms"] = ""
+        params["start_vm"] = "yes"
+        params["cdrom_unattended"] = ""
+        params["kernel"] = ""
+        params["initrd"] = ""
+        params["boot_once"] = "c"
+        params["kernel_params"] = ""
+        env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+        vm = env.get_vm(params.get("main_vm"))
```
Comment (XueqiangWei) [inline qemu/tests/qsd_install.py:42]: Suggest using the following two statements. I think no need to import env_process. 

vm.create(params=params)
vm.verify_alive()

with the update, the result was passed. I think we can remove line 4, 41 and 42. If I was wrong, please correct me. Thanks.
(1/1) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (662.11 s)

**Comment by menli820** [comment]
We have updated kar internal cfg for Windows image size as below per Microsoft document requirements.
should we also adjust the image size base on this?
Win11 - 64G
Win2022 - 36G
Win2012, Win10, Win2016, Win2019 - 32G

**Comment by menli820** [comment]
hit error when install the  win2016 guest, paste the log in the automation bug.

File: qemu/tests/qsd_install.py (hunk 2 above)
Comment (qingwangrh) [reply qemu/tests/qsd_install.py]: updated

File: qemu/tests/cfg/qsd_install.cfg (hunk 1 above)
Comment (qingwangrh) [reply qemu/tests/cfg/qsd_install.cfg]: So far, the main test is focused on x86, other platform is optional.
I will remove the limit in cfg.

**Comment by qingwangrh** [comment]
> hit error when install the win2016 guest, paste the log in the automation bug.

could you please try --customsparams="cpu_model_flags = 'vmx,+kvm_pv_unhalt'" due to Bug 2148399 - [QSD] get vhost_set_mem_table error when install windows os on vhost-user-blk-pci device

**Comment by XueqiangWei** [comment]
@qingwangrh Test results were passed. By the way, please update the previous comment, windows fails due to product bug. Add the detailed product bug information to automation bug. Thanks.

(01/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-
qemu.qsd_install: PASS (691.25 s)
(02/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-
qemu.qsd_install.q35: PASS (690.36 s)
(03/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qem
u.qsd_install: PASS (1051.53 s)
(04/12) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qem
u.qsd_install.q35: PASS (996.11 s)
(05/12) Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qe
mu.qsd_install: PASS (686.96 s)
(06/12) Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (684.76 s)
(07/12) Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install: PASS (907.79 s)
(08/12) Host_RHEL.m9.u2.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (895.06 s)
(09/12) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (760.36 s)
(10/12) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (989.66 s)
(11/12) Host_RHEL.m9.u2.ovmf.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.8.0.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (750.96 s)
(12/12) Host_RHEL.m9.u2.ovmf.raw.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (980.21 s)

File: qemu/tests/cfg/qsd_install.cfg [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    ovmf:
+        no i440fx
+        restore_ovmf_vars = yes
+        Windows:
+            send_key_at_install = ret
+        images = "stg0"
+        boot_drive_stg0 = no
+        force_create_image_stg0 = yes
+        image_name_stg0 = images/stg0
+        image_size_stg0 = 1G
```
Comment (menli820) [inline qemu/tests/cfg/qsd_install.cfg:40]: just a confirmation, for stg0, it only used for ovmf?

**Review by XueqiangWei — DISMISSED** [review]

**Comment by XueqiangWei** [comment]
LGTM,
Acked-by: xuwei@redhat.com

**Comment by zhenyzha** [comment]
Currently, ARM has not enabled qemu-kvm to support vhost-user-blk, so maybe only x86 or no aarch64 should be set in the configuration of the case.

@fbq815 @maxujun Could you check it out on your own platform? Prevent it from affecting your platform when it joins the block test loop.

CC: @PaulYuuu

**Comment by menli820** [comment]
LGTM

(2/2) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (671.79 s)

(2/2) Host_RHEL.m9.u2.ovmf.qcow2.virtio_blk.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.qsd_install.q35: PASS (1127.29 s)

**Comment by qingwangrh** [comment]
> Currently, ARM has not enabled qemu-kvm to support vhost-user-blk, so maybe only x86 or no aarch64 should be set in the configuration of the case.
> 
> @fbq815 @maxujun Could you check it out on your own platform? Prevent it from affecting your platform when it joins the block test loop.
> 
> CC: @PaulYuuu

@zhenyzha add "no aarch64" in cfg

**Comment by fbq815** [comment]
[root@l42 kar]# /usr/libexec/qemu-kvm -device vhost-user-blk-ccw,? 
qemu-kvm: -device vhost-user-blk-ccw,?: Device 'vhost-user-blk-ccw' not found
Seems s390x don't have this device either, so please add "no s390x" in the configuration file

**Comment by qingwangrh** [comment]
> [root@l42 kar]# /usr/libexec/qemu-kvm -device vhost-user-blk-ccw,? qemu-kvm: -device vhost-user-blk-ccw,?: Device 'vhost-user-blk-ccw' not found Seems s390x don't have this device either, so please add "no s390x" in the configuration file

Updated,thanks.

**Comment by maxujun** [comment]
vhost-user-blk-pci is not supported on ppc.

**Comment by XueqiangWei** [comment]
@YongxueHong Please help take a look at it. Many thanks.

File: qemu/tests/qsd_install.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        session.cmd(params["guest_cmd"])
+        logger.debug("Exit guest ...")
+        vm.destroy()
+        qsd.stop_daemon()
+        qsd = None
+    finally:
+        if qsd:
+            qsd.stop_daemon()
+
+    logger.info("Test Over")
```
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/qsd_install.py:53]: updated

**Review by YongxueHong — APPROVED** [review]
LGTM

**Review by XueqiangWei — APPROVED** [review]

**Review by zhenyzha — APPROVED** [review]

**Pull request merged**